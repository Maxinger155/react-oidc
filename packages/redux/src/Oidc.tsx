import React, { ComponentType, FC, PropsWithChildren } from 'react';
import { OidcProvider, loadUser } from 'redux-oidc';
import PropTypes from 'prop-types';
import {
  OidcRoutes,
  authenticationService,
  getUserManager,
  configurationPropTypes,
  UserStoreType,
} from '@3rdparty/react-oidc-core';
import AuthenticationCallback from './AuthenticationCallback';

const propTypes = {
  notAuthenticated: PropTypes.elementType,
  notAuthorized: PropTypes.elementType,
  callbackComponentOverride: PropTypes.elementType,
  sessionLostComponent: PropTypes.elementType,
  // eslint-disable-next-line react/require-default-props
  configuration: configurationPropTypes,
  store: PropTypes.object.isRequired,
  isEnabled: PropTypes.bool,
  children: PropTypes.node,
  UserStore: PropTypes.func,
};

const defaultPropsObject: Partial<OidcBaseProps> = {
  notAuthenticated: null,
  notAuthorized: null,
  callbackComponentOverride: null,
  sessionLostComponent: null,
  isEnabled: true,
  children: null,
  UserStore: null,
};

const withComponentOverrideProps = (Component: ComponentType, customProps: any) => (props: any) => (
  <Component callbackComponentOverride={customProps} {...props} />
);

export const OidcBase3rdparty = (props: any) => {
  const {
    isEnabled,
    children,
    store,
    callbackComponentOverride,
    configuration,
    notAuthenticated,
    notAuthorized,
    sessionLostComponent,
    UserStore,
    loadUser3rdparty,
    authenticationService3rdparty,
  } = props;

  const [ready, setReady] = React.useState(false);

  React.useEffect(() => {
    if (isEnabled) {
      const userManager = authenticationService3rdparty(configuration, UserStore);
      loadUser3rdparty(store, userManager);
      setReady(true);
    }
  }, [UserStore, configuration, isEnabled, store, loadUser3rdparty, authenticationService3rdparty]);

  return ready ? (
    <OidcProvider store={store} userManager={getUserManager()}>
      <OidcRoutes
        configuration={configuration}
        notAuthenticated={notAuthenticated}
        notAuthorized={notAuthorized}
        sessionLost={sessionLostComponent}
        callbackComponent={withComponentOverrideProps(
          AuthenticationCallback,
          callbackComponentOverride
        )}
      >
        {children}
      </OidcRoutes>
    </OidcProvider>
  ) : (
    <>{children}</>
  );
};

OidcBase3rdparty.propTypes = {
  ...propTypes,
  loadUser3rdparty: PropTypes.func.isRequired,
  authenticationService3rdparty: PropTypes.func.isRequired,
};

type OidcBaseProps = PropsWithChildren<{
  notAuthenticated?: ComponentType | null;
  notAuthorized?: ComponentType | null;
  callbackComponentOverride?: ComponentType | null;
  sessionLostComponent?: ComponentType | null;
  configuration: any;
  store: any;
  isEnabled: boolean;
  UserStore: UserStoreType;
}>;

const OidcBase: FC<OidcBaseProps> = props => (
  <OidcBase3rdparty
    loadUser3rdparty={loadUser}
    authenticationService3rdparty={authenticationService}
    {...props}
  />
);

// @ts-ignore
OidcBase.propTypes = propTypes;
OidcBase.defaultProps = defaultPropsObject;

export default OidcBase;
