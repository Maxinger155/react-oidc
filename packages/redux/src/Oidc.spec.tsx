// Link.react.test.js
import React from 'react';
import { render } from '@testing-library/react';
import { OidcBase3rdparty } from './Oidc';

jest.mock('redux-oidc', () => ({
  OidcProvider: jest.fn(() => <p>Render Something</p>),
}));

describe('redux.Oidc', () => {
  const loadUserMock = jest.fn();
  const authenticationServiceMock = jest.fn(() => 'userManager');
  const noopMock = () => {};

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('Render <OidcBase3rdparty/> correctly', () => {
    const configuration = {
      client_id: 'CSk26fuOE2NjQr17oCI1bKzBch9eUzF0',
      redirect_uri: 'http://localhost:3000/authentication/callback',
      response_type: 'id_token token',
      scope: 'openid profile email',
      authority: 'https://samplesreact.eu.auth0.com',
      silent_redirect_uri: 'http://localhost:3000/authentication/silent_callback',
      automaticSilentRenew: true,
      loadUserInfo: true,
      triggerAuthFlow: true,
    };
    const { asFragment } = render(
      <OidcBase3rdparty
        isEnabled={false}
        store={{ store: 'storeMock' }}
        configuration={configuration}
        loadUser3rdparty={loadUserMock}
        authenticationService3rdparty={authenticationServiceMock}
        UserStore={noopMock}
      >
        <p>isEnabled</p>
      </OidcBase3rdparty>
    );

    expect(asFragment()).toMatchSnapshot();
    expect(authenticationServiceMock).not.toHaveBeenCalled();
    expect(loadUserMock).not.toHaveBeenCalled();
  });

  it('Render <OidcBase3rdparty/> correctly when is enabled', () => {
    const configuration = {
      client_id: 'CSk26fuOE2NjQr17oCI1bKzBch9eUzF0',
      redirect_uri: 'http://localhost:3000/authentication/callback',
      response_type: 'id_token token',
      scope: 'openid profile email',
      authority: 'https://samplesreact.eu.auth0.com',
      silent_redirect_uri: 'http://localhost:3000/authentication/silent_callback',
      automaticSilentRenew: true,
      loadUserInfo: true,
      triggerAuthFlow: true,
    };
    const { asFragment } = render(
      <OidcBase3rdparty
        isEnabled
        store={{ store: 'storeMock' }}
        configuration={configuration}
        loadUser3rdparty={loadUserMock}
        authenticationService3rdparty={authenticationServiceMock}
        UserStore={noopMock}
      >
        <p>isEnabled</p>
      </OidcBase3rdparty>
    );
    expect(asFragment()).toMatchSnapshot();
    expect(authenticationServiceMock).toHaveBeenCalledWith(
      {
        authority: 'https://samplesreact.eu.auth0.com',
        automaticSilentRenew: true,
        client_id: 'CSk26fuOE2NjQr17oCI1bKzBch9eUzF0',
        loadUserInfo: true,
        redirect_uri: 'http://localhost:3000/authentication/callback',
        response_type: 'id_token token',
        scope: 'openid profile email',
        silent_redirect_uri: 'http://localhost:3000/authentication/silent_callback',
        triggerAuthFlow: true,
      },
      noopMock
    );
    expect(loadUserMock).toHaveBeenCalledWith({ store: 'storeMock' }, 'userManager');
  });
});
