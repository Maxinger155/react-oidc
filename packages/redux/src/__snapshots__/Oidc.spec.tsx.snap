// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux.Oidc Render <OidcBase3rdparty/> correctly 1`] = `
<DocumentFragment>
  <p>
    isEnabled
  </p>
</DocumentFragment>
`;

exports[`redux.Oidc Render <OidcBase3rdparty/> correctly when is enabled 1`] = `
<DocumentFragment>
  <p>
    Render Something
  </p>
</DocumentFragment>
`;
